src/frontend/Deprecation_analysis.ml

```diff
@@ -137,6 +137,52 @@ let rec collect_deprecated_expr (acc : (Location_span.t * string) list)
                 ) ]
           | _ -> [] ) in
       acc @ w @ List.concat_map l ~f:(fun e -> collect_deprecated_expr [] e)
+  | PrefixOp (PNot, ({emeta= {type_= UReal; loc; _}; _} as e)) ->
+      let acc =
+        acc
+        @ [ ( loc
+            , "Using a real as a boolean value is deprecated and will be \
+               disallowed in Stan 2.34. Use an explicit != 0 comparison \
+               instead. This can be automatically changed using the \
+               canonicalize flag for stanc" ) ] in
+      collect_deprecated_expr acc e
+  | BinOp (({emeta= {type_= UReal; loc; _}; _} as e1), (And | Or), e2)
+   |BinOp (e1, (And | Or), ({emeta= {type_= UReal; loc; _}; _} as e2)) ->
+      let acc =
+        acc
+        @ [ ( loc
+            , "Using a real as a boolean value is deprecated and will be \
+               disallowed in Stan 2.34. Use an explicit != 0 comparison \
+               instead. This can be automatically changed using the \
+               canonicalize flag for stanc" ) ] in
+      let acc = collect_deprecated_expr acc e1 in
+      let acc = collect_deprecated_expr acc e2 in
+      acc
+  | BinOp (({expr= BinOp (e1, op1, e2); emeta= {loc; _}} as e), op2, e3)
+    when Operator.(is_cmp op1 && is_cmp op2) ->
+      let pp_e = Pretty_printing.pp_typed_expression in
+      let pp = Operator.pp in
+      let acc =
+        acc
+        @ [ ( loc
+            , Fmt.str
+                "Found %a. This is interpreted as %a. Consider if the intended \
+                 meaning was %a but if it wasn't Stan 2.34 will require \
+                 explicit parenthesis. This can be automatically changed using \
+                 the canonicalize flag for stanc"
```

Comment:
> leave it a perpetual warning but never remove the syntax.

I think for now we should plan on doing this. We can always turn it into a removal warning later, but for now I personally wouldn't want it removed

I sort of see it as the typechecker warnings are useful for things like matrix-pow since they're not really deprecations, but rather type-specific warnings about unexpected or error-prone behavior. We could move them all elsewhere (which would also simplify the return type of the typechecker), but we'd want to either put them somewhere besides the module called `Deprecation_analysis` or rename that module to something more generic.